ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -312,99 +392,135 @@ runChainSync securityParam (ClientUpdates clientUpdates)
-             result <-
-               runPipelinedPeer protocolTracer codecChainSyncId clientChannel $
-                 chainSyncClientPeerPipelined $ client varCandidate
-             atomically $ writeTVar varClientResult (Just (Right result))
-             return ()
-        `catchAlsoLinked` \ex -> do
-          atomically $ writeTVar varClientResult (Just (Left ex))
-          -- Rethrow, but it will be ignored anyway.
-          throwIO ex
-      void $ forkLinkedThread registry "ChainSyncServer" $
-        runPeer nullTracer codecChainSyncId serverChannel
-                (chainSyncServerPeer server)
-
-    LogicalClock.waitUntilDone clock
-    -- Wait a random amount of time after the final tick for the chain sync
-    -- to finish
-    threadDelay 2000
+    --
+    -- Happens /immediately after/ the chain and clock effects schedule for
+    -- 'startSyncingAt'.
+    let initiateChainSync = do
+            (clientChannel, serverChannel) <- createConnectedChannels
+            -- Don't link the thread (which will cause the exception to be
+            -- rethrown in the main thread), just catch the exception and store
+            -- it, because we want a "regular ending".
+            void $ forkThread registry "ChainSyncClient" $
+              bracketChainSyncClient
+                 chainSyncTracer
+                 chainDbView
+                 varCandidates
+                 serverId
+                 maxBound $ \varCandidate -> do
+                   atomically $ modifyTVar varFinalCandidates $
+                     Map.insert serverId varCandidate
+                   result <-
+                     runPipelinedPeer protocolTracer codecChainSyncId clientChannel $
+                       chainSyncClientPeerPipelined $ client varCandidate
+                   atomically $ writeTVar varClientResult (Just (Right result))
+                   return ()
+              `catchAlsoLinked` \ex -> do
+                atomically $ writeTVar varClientResult (Just (Left ex))
+                -- Rethrow, but it will be ignored anyway.
+                throwIO ex
+            void $ forkLinkedThread registry "ChainSyncServer" $
+              runPeer nullTracer codecChainSyncId serverChannel
+                      (chainSyncServerPeer server)
+
+    -- If the candidate's tip's slot is ahead of the wall-clock, then the
+    -- ChainSync client mishandled a block from the future.
+    let checkTipTime :: m ()
+        checkTipTime = do
+            (now, candidates) <- atomically $ do
+              now        <- readTVar varLocalClock
+              candidates <- readTVar varCandidates >>= traverse readTVar
+              pure (now, candidates)
+            forM_ candidates $ \candidate -> case AF.headSlot candidate of
+              Origin  -> pure ()
+              At slot ->
+                when (now < toSkewedOnset slot) $ do
+                  error $ "future tip! " <> show (slot, now, candidate)
```

Comment:
Hmm, this check does not seem to fail when eg commenting out the `threadDelay` in `InFutureCheck`. I think this is due to using `readTVar varLocalClock`, which is completely unaffected by whether the ChainSync client calls `threadDelay` or not.

---

 - I find it plausible that this check is actually detecting all violations of the "Chronos principle", but eg only because `slotNo == blockNo` in this test, so it is eg not possible that during a logical tick, an client with an incorrect delay logic could add a header `H` with a very high slot number to the fragment, only to roll back to say `H'` with a sufficiently small one (which would be missed by this check). Maybe mention this, or say that this check isn't even intended to *fully* check the Chronos principle, but rather just a sanity check?
   
   Alternatives to the approach here are ofc also possible (probably a bit less lightweight), eg a new trace event that is emitted on header arrival, or some watcher of the candidate fragment that wakes up on every change to it (relying on io-sim that we don't miss any updates).
 - Due to the use of `error`, the QuickCheck counterexample information is missing on failure. This might be fine if we are confident that this check failing will always be trivial to fix without any counterexample context.

Reply:
Thanks for the note. I've made some changes and it does not fail if I inject that bug.

Primary changes:

- I replaced `varLocalClock` with merely the io-sim wall clock. This seems less error prone but does require a bit more ceremony in the setup.
- I replaced `toSkewedOnset` with `toOnset` in that conditional.
- The invariant violation is no longer handled via an exception; so the `counterexample`s et all work.
- Found a cheeky bug in the infrastructure: if the local chain gets ahead of the clock (due to clock skew), and then the peer rolls back to that point, the invariant can be violated, since the intersection is now ahead of the wall clock. Hence `mkClockUpdates` now distinguishes the greatest-slots so that the clock updating logic can handle them separately: exclude skew when updating for new max client slot but include it when updating for a new max server slot.

I'm OK with rollbacks being missed, since a separate QC counterexample wouldn't hide the violation in that way. I'll add a comment in the code.

Reply:
Looks good, nice that this also revealed the fact that we should not set the clock in such a way that we consider our own adopted blocks to be from the (near) future (I guess one *could* frame this as modeling clock rewind, but I don't think that is too useful).

> I'm OK with rollbacks being missed, since a separate QC counterexample wouldn't hide the violation in that way.

It felt *slightly* weird to implement a non-exhaustive check when it wouldn't be that hard to implement, but I agree it is fine, especially as it seems quite hard to implement the Chronos property incorrectly in such a way that this code even make a difference.